OpenRA.Game/Server/Server.cs

```diff
@@ -359,50 +375,127 @@ void ValidateClient(Connection newConn, string data)
 
-				Log.Write("server", "Client {0}: Accepted connection from {1}.",
-					newConn.PlayerIndex, newConn.Socket.RemoteEndPoint);
+					if (client.Fingerprint != null)
+						Log.Write("server", "Client {0}: Player fingerprint is {1}.",
+							newConn.PlayerIndex, client.Fingerprint);
 
-				foreach (var t in serverTraits.WithInterface<IClientJoined>())
-					t.ClientJoined(this, newConn);
+					foreach (var t in serverTraits.WithInterface<IClientJoined>())
+						t.ClientJoined(this, newConn);
 
-				SyncLobbyInfo();
+					SyncLobbyInfo();
 
-				Log.Write("server", "{0} ({1}) has joined the game.",
-					client.Name, newConn.Socket.RemoteEndPoint);
+					Log.Write("server", "{0} ({1}) has joined the game.",
+						client.Name, newConn.Socket.RemoteEndPoint);
+
+					// Report to all other players
+					SendMessage("{0} has joined the game.".F(client.Name), newConn);
+
+					// Send initial ping
+					SendOrderTo(newConn, "Ping", Game.RunTime.ToString(CultureInfo.InvariantCulture));
+
+					if (Dedicated)
+					{
+						var motdFile = Platform.ResolvePath(Platform.SupportDirPrefix, "motd.txt");
+						if (!File.Exists(motdFile))
+							File.WriteAllText(motdFile, "Welcome, have fun and good luck!");
+
+						var motd = File.ReadAllText(motdFile);
+						if (!string.IsNullOrEmpty(motd))
+							SendOrderTo(newConn, "Message", motd);
+					}
 
-				// Report to all other players
-				SendMessage("{0} has joined the game.".F(client.Name), newConn);
+					if (Map.DefinesUnsafeCustomRules)
+						SendOrderTo(newConn, "Message", "This map contains custom rules. Game experience may change.");
 
-				// Send initial ping
-				SendOrderTo(newConn, "Ping", Game.RunTime.ToString(CultureInfo.InvariantCulture));
+					if (!LobbyInfo.GlobalSettings.EnableSingleplayer)
+						SendOrderTo(newConn, "Message", TwoHumansRequiredText);
+					else if (Map.Players.Players.Where(p => p.Value.Playable).All(p => !p.Value.AllowBots))
+						SendOrderTo(newConn, "Message", "Bots have been disabled on this map.");
+				};
 
-				if (Dedicated)
+				if (!string.IsNullOrEmpty(handshake.Fingerprint) && !string.IsNullOrEmpty(handshake.AuthSignature))
 				{
-					var motdFile = Platform.ResolvePath(Platform.SupportDirPrefix, "motd.txt");
-					if (!File.Exists(motdFile))
-						File.WriteAllText(motdFile, "Welcome, have fun and good luck!");
+					waitingForAuthenticationCallback++;
 
-					var motd = File.ReadAllText(motdFile);
-					if (!string.IsNullOrEmpty(motd))
-						SendOrderTo(newConn, "Message", motd);
-				}
+					Action<DownloadDataCompletedEventArgs> onQueryComplete = i =>
+					{
+						PlayerProfile profile = null;
 
-				if (Map.DefinesUnsafeCustomRules)
-					SendOrderTo(newConn, "Message", "This map contains custom rules. Game experience may change.");
+						if (i.Error == null)
+						{
+							try
+							{
+								var yaml = MiniYaml.FromString(Encoding.UTF8.GetString(i.Result)).First();
+								if (yaml.Key == "Player")
```

Comment:
Do we want logging in case this condition is not met? (Or if `i.Error != null`?)